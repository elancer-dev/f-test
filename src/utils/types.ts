export type YMapsEvent = {
    get: (name: string) => {
        displayName: string;
        hl: [];
        type: string;
        value: string;
    };
}

export type TPoint = {
    coord: [number, number];
    name: string;
};

export type TState = {
    points: Array<TPoint>;
}

export type TAction<H extends object, D> = {
    handler: H;
    type: keyof H;
    data: D;
}

export type TDataArgumentType<F> = F extends (s: TState, d: infer D) => TState ? D : never;

export type TActionFunction = <H extends object, T extends keyof H>(handler: H, type: T, data: TDataArgumentType<H[T]>) => TAction<H, TDataArgumentType<H[T]>>;