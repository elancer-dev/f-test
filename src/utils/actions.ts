import { TState, TAction } from './types';

export const actionsHandler = {

    addPoint: (state: TState, data: { coord: [number, number]; name: string; }): TState => {

        var newState = { ...state };

        newState.points = [...state.points];

        newState.points.push(data);

        return newState;

    },

    updatePoint: (state: TState, data: { index: number; coord: [number, number]; name: string; }): TState => {

        var newState = { ...state };

        newState.points = [...state.points];

        newState.points[data.index].coord = data.coord;
        newState.points[data.index].name = data.name;

        return newState;

    },

    removePoint: (state: TState, data: { index: number }): TState => {

        var newState = { ...state };

        newState.points = [...state.points];

        newState.points.splice(data.index, 1);

        return newState;

    },

    changeOrder: (state: TState, data: { from: number; to: number; }): TState => {

        var newState = { ...state };

        newState.points = [...state.points];

        const item = newState.points.splice(data.from, 1);

        newState.points.splice(data.to, 0, item[0])

        return newState;

    },

}

export const action = <H extends object, T extends keyof H, D>(h: H, t: T, d: D): TAction<H, D> => {
    return { type: t, data: d, handler: h }
}