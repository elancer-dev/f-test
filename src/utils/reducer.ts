import { TState, TAction } from './types';

export const initialState: TState = {
    points: [
        // { coord: [55.753215, 37.622504], name: 'Москва', },
        // { coord: [56.185102, 36.977631], name: 'Солнечногорск', },
        // { coord: [55.991390, 36.484994], name: 'село Новопетровское', },
    ],
}

export function reducer(state: TState = initialState, action: TAction<any, any>): TState {

    if (action.handler !== undefined && typeof action.handler[action.type] === 'function') {
        return action.handler[action.type](state, action.data);
    } else {
        return state;
    }

}