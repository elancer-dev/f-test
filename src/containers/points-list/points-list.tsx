import React, { DragEvent } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { TState, TActionFunction, TPoint } from '../../utils/types';
import { action, actionsHandler } from './../../utils/actions';
import PointsListView from './../../components/points-list-view/points-list-view';

type TProps = {}

type TPS = TProps & {
    points: Array<TPoint>;
    action: TActionFunction;
}

export class PointsList extends React.PureComponent<TPS> {

    private currentItem: any = null;
    private offset = 0;

    dragStart = (e: DragEvent<HTMLDivElement>) => {

        e.dataTransfer.effectAllowed = "move";
        e.dataTransfer.setData('text/html', 'data');
        this.currentItem = e.currentTarget;

    }

    dragOver = (e: DragEvent<HTMLDivElement>) => {

        var order = this.currentItem.style.order;
        this.currentItem.style.order = e.currentTarget.style.order;
        e.currentTarget.style.order = order;

    }

    dragEnd = (e: DragEvent<HTMLDivElement>) => {

        var from = parseInt(this.currentItem.getAttribute('data-order') + '');
        var to = parseInt(this.currentItem.style.order + '') - this.offset

        this.props.action(actionsHandler, 'changeOrder', { from, to });

        this.currentItem = null;

    }

    removeItem = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {

        if (e.currentTarget.parentElement) {
            var index = parseInt(e.currentTarget.parentElement.getAttribute('data-order') + '');
            this.props.action(actionsHandler, 'removePoint', { index });
        }

    }

    render = () => {

        this.offset += this.props.points.length;

        return (
            <PointsListView points={this.props.points} offset={this.offset} dragStart={this.dragStart} dragOver={this.dragOver} dragEnd={this.dragEnd} removeItem={this.removeItem} />
        )

    }

}

const mapStateToProps = (state: TState) => {

    return {
        points: state.points,
    };

};

const mapDispatchToProps = (dispatch: Dispatch) => {

    return bindActionCreators({ action: action }, dispatch);

};

export default connect(mapStateToProps, mapDispatchToProps)(PointsList);