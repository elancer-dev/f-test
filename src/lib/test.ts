import { TState } from "../utils/types";

export const getState = (): TState => {
    return {
        points: [
            { coord: [0, 0], name: 'name1' },
            { coord: [1, 1], name: 'name2' },
            { coord: [2, 2], name: 'name3' },
            { coord: [3, 3], name: 'name4' },
        ]
    }
}