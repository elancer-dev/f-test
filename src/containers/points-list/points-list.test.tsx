import React, { Component } from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import TestRenderer from 'react-test-renderer';
import { actionsHandler } from "./../../utils/actions";
import { getState } from "./../../lib/test";
import { PointsList } from './points-list';

export default describe('<PointsList />', () => {

    it('Renders PointsList component', () => {

        const action = jest.fn();
        const state = getState();
        const pointsArrayLength = state.points.length;

        const root = <PointsList action={action} points={state.points} />;

        const component = ReactTestUtils.renderIntoDocument<Component>(root);

        const tree = TestRenderer.create(root).toJSON();
        expect(tree).toMatchSnapshot();

        const listItems = ReactTestUtils.scryRenderedDOMComponentsWithClass((component as Component), 'points-list__item-remove');

        ReactTestUtils.Simulate.click(listItems[listItems.length - 1]);

        expect(action).toHaveBeenCalledWith(actionsHandler, 'removePoint', { index: pointsArrayLength - 1 });

    });

});