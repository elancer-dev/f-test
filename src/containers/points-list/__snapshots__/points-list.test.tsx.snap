// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PointsList /> Renders PointsList component 1`] = `
<div
  className="points-list"
>
  <div
    className="points-list__item"
    data-order={0}
    draggable={true}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    style={
      Object {
        "order": 4,
      }
    }
  >
    <div
      className="points-list__item-index"
    >
      1
    </div>
    <div
      className="points-list__item-name"
    >
      name1
    </div>
    <div
      className="points-list__item-remove"
      onClick={[Function]}
    >
      x
    </div>
  </div>
  <div
    className="points-list__item"
    data-order={1}
    draggable={true}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    style={
      Object {
        "order": 5,
      }
    }
  >
    <div
      className="points-list__item-index"
    >
      2
    </div>
    <div
      className="points-list__item-name"
    >
      name2
    </div>
    <div
      className="points-list__item-remove"
      onClick={[Function]}
    >
      x
    </div>
  </div>
  <div
    className="points-list__item"
    data-order={2}
    draggable={true}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    style={
      Object {
        "order": 6,
      }
    }
  >
    <div
      className="points-list__item-index"
    >
      3
    </div>
    <div
      className="points-list__item-name"
    >
      name3
    </div>
    <div
      className="points-list__item-remove"
      onClick={[Function]}
    >
      x
    </div>
  </div>
  <div
    className="points-list__item"
    data-order={3}
    draggable={true}
    onDragEnd={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    style={
      Object {
        "order": 7,
      }
    }
  >
    <div
      className="points-list__item-index"
    >
      4
    </div>
    <div
      className="points-list__item-name"
    >
      name4
    </div>
    <div
      className="points-list__item-remove"
      onClick={[Function]}
    >
      x
    </div>
  </div>
</div>
`;
